packages/table/src/table.tsx

```diff
@@ -681,28 +681,46 @@ export class Table extends AbstractComponent<ITableProps, ITableState> {
             selectionModes,
         } = nextProps;
 
-        const newChildArray = React.Children.toArray(children) as Array<React.ReactElement<IColumnProps>>;
+        const hasNewChildren = this.props.children !== nextProps.children;
+        const newChildArray = hasNewChildren
+            ? (React.Children.toArray(children) as Array<React.ReactElement<IColumnProps>>)
+            : this.childrenArray;
         const numCols = newChildArray.length;
 
-        // Try to maintain widths of columns by looking up the width of the
-        // column that had the same `ID` prop. If none is found, use the
-        // previous width at the same index.
-        const previousColumnWidths = newChildArray.map((child: React.ReactElement<IColumnProps>, index: number) => {
-            const mappedIndex = this.columnIdToIndex[child.props.id];
-            return this.state.columnWidths[mappedIndex != null ? mappedIndex : index];
-        });
-
-        // Make sure the width/height arrays have the correct length, but keep
-        // as many existing widths/heights when possible. Also, apply the
-        // sparse width/heights from props.
+        let gridInvalidationRequired = false;
         let newColumnWidths = this.state.columnWidths;
-        newColumnWidths = Utils.arrayOfLength(newColumnWidths, numCols, defaultColumnWidth);
-        newColumnWidths = Utils.assignSparseValues(newColumnWidths, previousColumnWidths);
-        newColumnWidths = Utils.assignSparseValues(newColumnWidths, columnWidths);
+        if (
+            defaultColumnWidth !== this.props.defaultColumnWidth ||
+            columnWidths !== this.props.columnWidths ||
+            (hasNewChildren && !this.areChildrenSameShape(newChildArray))
+        ) {
+            // Try to maintain widths of columns by looking up the width of the
+            // column that had the same `ID` prop. If none is found, use the
+            // previous width at the same index.
+            const previousColumnWidths = newChildArray.map((child: React.ReactElement<IColumnProps>, index: number) => {
+                const mappedIndex = this.columnIdToIndex[child.props.id];
+                return this.state.columnWidths[mappedIndex != null ? mappedIndex : index];
+            });
+
+            // Make sure the width/height arrays have the correct length, but keep
+            // as many existing widths/heights when possible. Also, apply the
+            // sparse width/heights from props.
+            newColumnWidths = Utils.arrayOfLength(newColumnWidths, numCols, defaultColumnWidth);
+            newColumnWidths = Utils.assignSparseValues(newColumnWidths, previousColumnWidths);
+            newColumnWidths = Utils.assignSparseValues(newColumnWidths, columnWidths);
+            gridInvalidationRequired = true;
```

Comment:
nit: this line is a separate "thought" unrelated to the previous commented chunk, so i'd put a newline above.